rust/geoarrow-array/src/cast.rs

```diff
@@ -254,6 +263,282 @@ impl AsGeoArrowArray for Arc<dyn GeoArrowArray> {
+    use GeoArrowType::*;
+    match arr.data_type() {
+        Point(_) => impl_to_wkb(arr.as_point()),
+        LineString(_) => impl_to_wkb(arr.as_line_string()),
+        Polygon(_) => impl_to_wkb(arr.as_polygon()),
+        MultiPoint(_) => impl_to_wkb(arr.as_multi_point()),
+        MultiLineString(_) => impl_to_wkb(arr.as_multi_line_string()),
+        MultiPolygon(_) => impl_to_wkb(arr.as_multi_polygon()),
+        Geometry(_) => impl_to_wkb(arr.as_geometry()),
+        GeometryCollection(_) => impl_to_wkb(arr.as_geometry_collection()),
+        Rect(_) => impl_to_wkb(arr.as_rect()),
+        Wkb(typ) => {
+            // Note that here O is the _target_ offset type
+            if O::IS_LARGE {
+                // We need to convert from i32 to i64
+                let large_arr: WkbArray<i64> = arr.as_wkb::<i32>().clone().into();
+                let array = large_arr.to_array_ref().as_binary::<O>().clone();
+                Ok(WkbArray::new(array, typ.metadata().clone()))
+            } else {
+                // Since O is already i32, we can just go via ArrayRef, and use .as_binary to cast
+                // to O
+                let array = arr.as_wkb::<i32>().to_array_ref();
+                let array = array.as_binary::<O>().clone();
+                Ok(WkbArray::new(array, typ.metadata().clone()))
+            }
+        }
+        LargeWkb(typ) => {
+            if O::IS_LARGE {
+                // Since O is already i64, we can just go via ArrayRef, and use .as_binary to cast
+                // to O
+                let array = arr.as_wkb::<i64>().to_array_ref();
+                let array = array.as_binary::<O>().clone();
+                Ok(WkbArray::new(array, typ.metadata().clone()))
+            } else {
+                // We need to convert from i64 to i32
+                let small_arr: WkbArray<i32> = arr.as_wkb::<i64>().clone().try_into()?;
+                let array = small_arr.to_array_ref().as_binary::<O>().clone();
+                Ok(WkbArray::new(array, typ.metadata().clone()))
+            }
+        }
+        Wkt(_) => impl_to_wkb(arr.as_wkt::<i32>()),
+        LargeWkt(_) => impl_to_wkb(arr.as_wkt::<i64>()),
+    }
+}
+
+fn impl_to_wkb<'a, O: OffsetSizeTrait>(geo_arr: &'a impl ArrayAccessor<'a>) -> Result<WkbArray<O>> {
+    let geoms = geo_arr
+        .iter()
+        .map(|x| x.transpose())
+        .collect::<Result<Vec<_>>>()?;
+    let wkb_type = WkbType::new(geo_arr.data_type().metadata().clone());
+    Ok(WkbBuilder::from_nullable_geometries(geoms.as_slice(), wkb_type).finish())
+}
+
+/// Parse a [WkbArray] to a [GeoArrowArray] with the designated [GeoArrowType].
+///
+/// Note that the GeoArrow metadata on the new array is taken from `to_type` **not** the original
+/// array. Ensure you construct the [GeoArrowType] with the correct metadata.
+pub fn from_wkb<O: OffsetSizeTrait>(
+    arr: &WkbArray<O>,
+    to_type: GeoArrowType,
+    prefer_multi: bool,
```

Comment:
@paleolimbot do you have an idea of how to make this API nicer? In particular, the `prefer_multi` flag says, for `GeometryArray` and `GeometryCollectionArray`, if you're adding a point, linestring, or polygon, should those be added to the e.g. `PointArray` or the `MultiPointArray`. The latter might be preferable if then you can "downcast" the GeometryArray to a single concrete array if you have a mix of `PointArray` and `MultiPointArray`.

But having `prefer_multi` as a flag here for all geometry types when it only matters for `GeometryArray` and `GeometryCollectionArray` is pretty ugly.